### PR TITLE
8310743: assert(reserved_rgn != nullptr) failed: Add committed region, No reserved region found

### PR DESCRIPTION
**Pull request by xmas92**
JDK-8309675 tried to solve the issue of NMT reporting to much committed memory by faking the virtual addresses registered when committing physical memory. This mapping was simply the `virtual address  = physical offset in backing file + heap base`. This works when ZGC gets a contiguous heap which starts at the heap base. However fails if it does not start at the heap base or is discontiguous.

This proposal fixes this by keeping track of all the reserved virtual address ranges and by creating a mapping from the physical offsets to the actual reserved virtual address ranges. If the physical offsets committed maps to multiple reserved ranges, it is processed as multiple commits.

The develop flag `ZForceDiscontiguousHeapReservations` is introduced to test the discontiguous heap feature and interactions.

Testing: 
 * Added test on all Oracle platforms
 * GHA
 * Running tier1-3

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8310743](https://bugs.openjdk.org/browse/JDK-8310743): assert(reserved_rgn != nullptr) failed: Add committed region, No reserved region found (**Bug** - P2)

### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14730/head:pull/14730` \
`$ git checkout pull/14730`

Update a local copy of the PR: \
`$ git checkout pull/14730` \
`$ git pull https://git.openjdk.org/jdk.git pull/14730/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14730`

View PR using the GUI difftool: \
`$ git pr show -t 14730`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14730.diff">https://git.openjdk.org/jdk/pull/14730.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14730#issuecomment-1614642326)

### CODE REVIEW

**Review by stefank — APPROVED** [review]

File: src/hotspot/share/gc/z/zNMT.cpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    left_to_process -= sub_range_size;
+    if (left_to_process == 0) {
+      // Processed all nmt registrations
+      return;
+    }
+
+    offset_in_reservation = 0;
+  }
+
+  assert(left_to_process == 0, "everything was not commited");
```
Comment (albertnetymk) [inline src/hotspot/share/gc/z/zNMT.cpp:91]: When will this be reachable? I'd expect the if-check inside the for-loop to handle exit-condition.

File: src/hotspot/share/gc/z/zNMT.cpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    const uintptr_t sub_range_addr = untype(reservation_start) + offset_in_reservation;
+
+    // commit / uncommit memory
+    if (commit) {
+      MemTracker::record_virtual_memory_commit((void*)sub_range_addr, sub_range_size, CALLER_PC);
+    } else {
+      if (MemTracker::enabled()) {
+        Tracker tracker(Tracker::uncommit);
+        tracker.record((address)sub_range_addr, sub_range_size);
+      }
```
Comment (albertnetymk) [inline src/hotspot/share/gc/z/zNMT.cpp:79]: Sth like `record_virtual_memory_uncommit` would make this more symmetric. (Better be done in a followup PR though.)

File: src/hotspot/share/gc/z/zNMT.cpp [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+size_t ZNMT::reservation_index(zoffset offset, size_t* offset_in_reservation) {
+  assert(_num_reservations > 0, "at least one reservation must exist");
+
+  size_t index = 0;
+  *offset_in_reservation = untype(offset);
+  for (; index < _num_reservations; ++index) {
+    const size_t reservation_size = _reservations[index]._size;
+    if (*offset_in_reservation < reservation_size) {
+      break;
```
[the remark this answers is not in the document]
Comment (xmas92) [reply src/hotspot/share/gc/z/zNMT.cpp:44]: It was purposely written like this avoiding a `fatal`/`guarantee`/`ShouldNotReachHere`, the intent is to have bugs only be caught in debug. If we return `_num_reservations` as an index all that happens is that we do not register some commit with nmt. Which seems more favourable than crashing if some bug were to slip through into production.

File: src/hotspot/share/gc/z/zNMT.cpp (hunk 1 above)
Comment (xmas92) [reply src/hotspot/share/gc/z/zNMT.cpp:91]: If we run out of reservations. That is the virtual memory reservations range is less than the physical memory offset we are committing. Such a thing should not be reachable and is a bug. But if it does happen it should be caught here.

File: src/hotspot/share/gc/z/zNMT.cpp (hunk 2 above)
Comment (xmas92) [reply src/hotspot/share/gc/z/zNMT.cpp:79]: I agree. Purposely kept this minimal as it needs to be backported to 21.

**Review by albertnetymk — APPROVED** [review]
I think it's better to fail-early when sth unexpected occurs, so both places warrant a should-not-reach-here; subjective though.

**Comment by xmas92** [comment]
I generally agree with the fail early strategy. My personal ILW assessment of the two approaches made me favour this one. But maybe I'll reassess this in the future. 

Thanks for all reviews.

/integrate

**Pull request closed without merging**